src/development/dev-tests-coding.md

```diff
@@ -0,0 +1,165 @@
+  If you were reading such a test, you'd probably skip all of the initialisation code and come back to it after having read the test cases themselves.
+* __No explicit intent__
+  As you read the test cases, you will need to go back to the shared data and keep in your head what user1 and user2 are being used for. How they are used is down to the test case. The lack of intent from the field definitions means the reader has to keep this context in their head.
+
+A better approach is to declare the fixture within the test.
+
+## Create fixture data in factory methods
+
+If we’ve decided to avoid shared fixture data, then won’t we end up with lots of object creation code within our tests? Yes, you can. This is ok if object creation can be done with clear intent in a one liner. But even using lombok builders, object creation can often span multiple lines which leads to needless boilerplate.
+
+A very simple solution to this is to use simple factory methods at the bottom of the test class.
+
+* __Less verbose__
+  Exposing a method that only takes the argument that varies for the test case means all the default boilerplate is handled in one place, out of the way.
+* __Clear intent__
+  Exposing multiple methods to describe different situations can clearly express why a fixture object is being created.
+* __Clear naming__
+  As objects are created within the test, they can be assigned to variables that suit the test case.
+
+Use methods with clear names and intent to create fixture data when it is needed.
+Name the created fixture objects to further clarify the intent within the test.
+
+## Testing with date and time
+
+When code is dependent on date or time, your tests must take control of time so they can test all the different flows and expectations.
+
+We don't hardcode fixture data to specific dates/times far in the future. This test will fail one day and another poor developer will have to come along and mop up the problem.
+
+There are two common approaches to testing with time:
+
+* __Offset input data__
+  Use system date and/or time and adjust with an offset to create values in fixture data. The code will apply logic based on the current system time and the input data. This is a simple approach that works well in most cases.
+
+* __Control system time__
+  Sometimes you can't use fixture data: For example, in the [Greeter kata](https://github.com/wix/tdd-katas), a salutation changes through the course of the day. We obviously can't write tests that only pass in the morning. Instead we need to control system time.
+  
+  Most languages have a way to control system time in code.
+  * [Java](https://www.baeldung.com/java-override-system-time)
+  * [Javascript using Jest](https://codewithhugo.com/mocking-the-current-date-in-jest-tests/)
+
+## Test one thing
+
+If you adopt Test Driven Development (TDD) you should find that your test cases are lean, with just enough fixture data to test a single additional requirement.
```

Comment:
Can this be expanded on to cover the importance of a variety of test data.  If you're writing a method that processes a string it might be interpreted that one or two tests coving sample strings is enough.  But tests should also cover empty strings, whitespace only strings or null.  Otherwise you'll find out in production how the method behaves with such values.

Reply:
I'm adding a section on _Test coverage_.

It explains that code coverage tools aren't everything and devs need to consider 'coverage' in a wider sense. I've pulled in exception handling, null/empty checking, empty collections, boundary conditions. Anything else you can think of, please comment when I've pushed.

Resolving this issue now, but I'll resolve the exception issue once I've pushed.